src/elements/play-assets/file-upload-dropper.ts

```diff
@@ -0,0 +1,280 @@
+import {
+  css,
+  html,
+  LitElement,
+  type PropertyValues,
+  type TemplateResult
+} from 'lit'
+import {customElement, property, state} from 'lit/decorators.js'
+import {classMap} from 'lit-html/directives/class-map.js'
+import {ifDefined} from 'lit-html/directives/if-defined.js'
+import {
+  fileAccessContext,
+  type FilePickerType,
+  flattenAcceptTypes,
+  hasFileAccessAPI,
+  tryGetAsFilesystemHandle
+} from '../../utils/file-access-api.js'
+import {cssReset} from '../../utils/css-reset.js'
+import {Bubble} from '../../utils/bubble.js'
+
+declare global {
+  interface HTMLElementEventMap {
+    'files-selected': CustomEvent<FileSelection>
+    cancelled: CustomEvent<never>
+  }
+  interface HTMLElementTagNameMap {
+    'file-upload-dropper': FileUploadDropper
+  }
+}
+
+export type FileSelection = {
+  // Provided in fallback environments where File Access API is not available
+  files?: File[]
+  // Provided in environments where File Access API is available
+  fileHandles?: FileSystemFileHandle[]
+}
+
+@customElement('file-upload-dropper')
+export class FileUploadDropper extends LitElement {
+  static override readonly styles = css`
+    ${cssReset}
+
+    input[type='file'] {
+      display: none;
+    }
+
+    #fileDrop {
+      border: 3px dashed black;
+      border-radius: 8px;
+      padding: var(--space);
+      display: flex;
+      flex-direction: column;
+      align-items: center;
+      cursor: pointer;
+    }
+
+    #fileDrop.dragenter {
+      border-color: blue;
+    }
+
+    #errorMessage {
+      color: red;
+      font-size: 0.9em;
+    }
+  `
+
+  @property({attribute: 'accept-types', type: Array, reflect: true})
+  acceptTypes: FilePickerType[] = []
+
+  @property({attribute: 'multiple', type: Boolean, reflect: true})
+  multiple: boolean = false
+
+  @state()
+  private _dragging: boolean = false
+
+  @state()
+  private _errorMessage: string | undefined
+
+  @state()
+  private _accept: string | undefined
+
+  #allowedMimes: string[] = []
+  #acceptPatterns: RegExp[] = []
+
+  protected override update(changedProperties: PropertyValues<this>) {
+    super.update(changedProperties)
+    if (changedProperties.has('acceptTypes')) {
+      this._accept = flattenAcceptTypes(this.acceptTypes)
```

Comment:
I think this will drive another state change. should we use willUpdate()?

Reply:
Modifying state in `update()` does not trigger further updates/renders: https://lit.dev/docs/api/ReactiveElement/#ReactiveElement.update  Though `willUpdate()` would be fine since it occurs before other callbacks, it's not necessary here since `_accept` isn't critical to rendering or other updates since it's for the file picker dialog.